### PR TITLE
:world_map: Add navigation presenter for iOS

### LINKED ISSUE #4857: Create a presenter for child view controllers
Depends on #4852

We need to create an IosPresenter that is capable of displaying ViewModels that are associated with child View Controllers

### PR DESCRIPTION
**Pull request by juanlaube**
## What's this?
Adds the navigation presenter for iOS to push view controllers on a navigation stack. This kind of presentation is what used to be a ChildPresentation in MvvmCross.

### Relationships

Closes #4857

## Why do we want this?
🇬🇧 ➡️ 🚪 

## How is it done?
It's a new presenter. One important detail is the new `Task Cancel()` method in `IViewModel` this is because pressing the back button in the navigation controller needs to cancel the `CloseCompletionSource`.

### Why not another way?
I thought about subclassing `UINavigationController` instead of overriding `DidMoveToParentViewController` to handle the close action, but went with this because it's exactly how MvvmCross handles that.

### Side effects
None.

## :squid: Permissions
🦑

### CODE REVIEW

File: Toggl.Core.UI/ViewModels/Base/ViewModel.cs [hunk 1]
```diff
@@ -18,6 +18,12 @@ public async Task Finish(TOutput output)
             CloseCompletionSource.SetResult(output);
         }
 
+        public async Task Cancel()
+        {
+            await View.Close();
+            CloseCompletionSource.SetCanceled();
```
Comment (heytherewill) [inline Toggl.Core.UI/ViewModels/Base/ViewModel.cs]: So elegant

File: Toggl.Core.UI/ViewModels/Base/ViewModel.cs [hunk 2]
```diff
@@ -46,6 +52,10 @@ public virtual void ViewDisappeared()
 
         public virtual void ViewDestroyed()
         {
+            if (typeof(TOutput) == typeof(Unit))
```
Comment (heytherewill) [inline Toggl.Core.UI/ViewModels/Base/ViewModel.cs]: I don't like this. The VM shouldn't cancel itself, specially not in a lifecycle method.

File: Toggl.iOS/Presentation/IosPresenter.cs [hunk 3]
```diff
@@ -36,5 +36,13 @@ public IosPresenter(UIWindow window, AppDelegate appDelegate)
 
             return tcs.Task;
         }
+
+        protected UIViewController FindPresentedViewController()
+            => FindPresentedViewController(Window.RootViewController);
+
+        private UIViewController FindPresentedViewController(UIViewController current)
```
Comment (heytherewill) [inline Toggl.iOS/Presentation/IosPresenter.cs]:
```suggestion
        private UIViewController findPresentedViewController(UIViewController current)
```

File: Toggl.iOS/Presentation/IosPresenter.cs [hunk 4]
```diff
@@ -36,5 +36,13 @@ public IosPresenter(UIWindow window, AppDelegate appDelegate)
...
+            => current.PresentedViewController == null || current.PresentedViewController.IsBeingDismissed
+                ? current
+                : FindPresentedViewController(current.PresentedViewController);
```
Comment (heytherewill) [inline Toggl.iOS/Presentation/IosPresenter.cs]:
```suggestion
                : findPresentedViewController(current.PresentedViewController);
```

File: Toggl.iOS/ViewControllers/Reactive/ReactiveViewController.cs [hunk 5]
```diff
@@ -51,6 +51,16 @@ public override void ViewDidDisappear(bool animated)
             ViewModel?.ViewDisappeared();
         }
 
+        public override void DidMoveToParentViewController(UIViewController parent)
+        {
+            base.DidMoveToParentViewController(parent);
+
+            if (parent == null)
```
Comment (heytherewill) [inline Toggl.iOS/ViewControllers/Reactive/ReactiveViewController.cs]: America, please explain

File: Toggl.Core.UI/ViewModels/Base/ViewModel.cs (hunk 1 above)
Comment (juanlaube) [reply Toggl.Core.UI/ViewModels/Base/ViewModel.cs]: :sogood:

File: Toggl.iOS/ViewControllers/Reactive/ReactiveTableViewController.cs [hunk 6]
```diff
@@ -51,6 +51,20 @@ public override void ViewDidDisappear(bool animated)
             ViewModel?.ViewDisappeared();
         }
 
+        public override void DidMoveToParentViewController(UIViewController parent)
+        {
+            base.DidMoveToParentViewController(parent);
+
+            // When this is removed from the parent controller (e.g: a pop action in a navigation controller)
+            // then the view is destroyed and the view model needs to be cancelled
+            if (parent == null)
```
[the remark this answers is not in the document]
Comment (juanlaube) [reply Toggl.iOS/ViewControllers/Reactive/ReactiveTableViewController.cs]: Fixed

**Pull request merged**